File: 244.shortest-word-distance-ii.java
import java.util.*;
/*
 * @lc app=leetcode id=244 lang=java
 *
 * [244] Shortest Word Distance II
 */

// @lc code=start
class WordDistance {
    private Map<String, List<Integer>> wordIdxs;
    private Map<String, Integer> cache;     // useful when a pair is queried multiple times

    public WordDistance(String[] wordsDict) {
        wordIdxs = new HashMap<>();
        cache = new HashMap<>();

        // Note: computing minimum distances between all pairs of words here
        // is too time-consuming.

        for (int i = 0; i < wordsDict.length; ++i) {
            wordIdxs.computeIfAbsent(wordsDict[i], k -> new ArrayList<>()).add(i);
        }
    }
    
    public int shortest(String word1, String word2) {
        String key = word1 + '_' + word2;
        if (cache.containsKey(key))  return cache.get(key);
        List<Integer> l1 = wordIdxs.get(word1), l2 = wordIdxs.get(word2);
        
        // l1 and l2 are sorted
        int i = 0, j = 0, ans = Integer.MAX_VALUE;
        while (i < l1.size() && j < l2.size()) {
            int idx1 = l1.get(i), idx2 = l2.get(j);
            if (idx1 > idx2) {
                ans = Math.min(ans, idx1 - idx2);
                ++j;
            } else {
                ans = Math.min(ans, idx2 - idx1);
                ++i;
            }

            // Early break: cannot get smaller than 1
            if (ans == 1) {
                cache.put(key, 1);
                return 1;
            }
        }
        cache.put(key, ans);
        return ans;
    }
}

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance obj = new WordDistance(wordsDict);
 * int param_1 = obj.shortest(word1,word2);
 */
// @lc code=end

File: 244.shortest-word-distance-ii.py
#
# @lc app=leetcode id=244 lang=python3
#
# [244] Shortest Word Distance II
#
from collections import defaultdict
from typing import List

# @lc code=start
class WordDistance:
    # T: O(n)
    # S: O(n)
    def __init__(self, wordsDict: List[str]):
        self.word_idxs = defaultdict(list)
        for i, w in enumerate(wordsDict):
            self.word_idxs[w].append(i)
        
        # Assume no more words will be added, can cache the query results
        # k: (w1, w2), v: shortest dist 
        self.cache = dict()

    # T: O(n)
    # S: O(1)
    def shortest(self, word1: str, word2: str) -> int:
        cache_k = (word1, word2)
        cache_v = self.cache.get(cache_k, -1)
        if cache_v != -1:
            return cache_v

        idxs1 = self.word_idxs[word1]
        idxs2 = self.word_idxs[word2]
        if len(idxs1) == 0 or len(idxs2) == 0:
            return -1
        ans = float('inf')
        i = 0
        j = 0
        while i < len(idxs1) and j < len(idxs2):
            idx1 = idxs1[i]
            idx2 = idxs2[j]
            if idx1 > idx2:
                ans = min(ans, idx1 - idx2)
                j += 1
            else:
                ans = min(ans, idx2 - idx1)
                i += 1
            
            if ans == 1:
                # minimum valid distance, stop early
                break
        
        self.cache[cache_k] = ans
        return ans


# Your WordDistance object will be instantiated and called as such:
# obj = WordDistance(wordsDict)
# param_1 = obj.shortest(word1,word2)
# @lc code=end

